src/check_presence.c

```diff
@@ -0,0 +1,533 @@
+  }
+
+  int rc = mdb_env_create(&env);
+  if (rc) {
+    ERROR(PLUGIN_NAME " plugin: mdb_env_create failed: %s (%d)",
+          mdb_strerror(rc), rc);
+    return -1;
+  }
+
+  rc = mdb_env_open(env, state_datastore, MDB_NOSUBDIR, 0664);
+  if (rc) {
+    ERROR(PLUGIN_NAME " plugin: opening path '%s' failed: %s (%d)",
+          state_datastore, mdb_strerror(rc), rc);
+    mdb_env_close(env);
+    env = NULL;
+    return -1;
+  }
+
+  return start_thread();
+}
+
+static int presence_shutdown(void) {
+
+  if (env) {
+
+    INFO(PLUGIN_NAME " plugin: shutting down thread.");
+    if (stop_thread() < 0)
+      return -1;
+
+    mdb_env_close(env);
+  }
+
+  return 0;
+}
+
+static int presence_write(const data_set_t *ds, const value_list_t *vl,
+                          __attribute__((unused)) user_data_t *user_data) {
+
+  /* step aside quitely if not configured */
+  if (!env) {
+    return 0;
+  }
+
+  struct timeval tv_now = {0};
+
+  int rc = gettimeofday(&tv_now, /* struct timezone = */ NULL);
+  if (rc != 0) {
+    ERROR(PLUGIN_NAME " plugin: gettimeofday failed: %s", STRERRNO);
+    return -1;
+  }
+
+  /*
+   * Fast path - have we seen this host before?
+   *
+   * WHile this step is not strictly necessary, in the vast majority of
+   * cases we will have seen the host before, and we will have seen the
+   * host many times in the same second as each metric is written.
+   *
+   * The LMDB key value store offers very cheap lock free reads, allowing
+   * multiple threads to handle writes without any mutexes.
+   *
+   * This first step discards unnecessary writes as quickly as possible.
+   */
+
+  MDB_val key = {0}, data = {0};
+
+  key.mv_size = sizeof(vl->host);
+  key.mv_data = (void *)vl->host;
+
+  MDB_txn *txn = NULL;
+
+  rc = mdb_txn_begin(env, NULL, MDB_RDONLY, &txn);
+  if (rc) {
+    ERROR(PLUGIN_NAME " plugin: mdb_txn_begin returned: %s (%d)",
+          mdb_strerror(rc), rc);
+    return -1;
+  }
+
+  MDB_dbi dbi = {0};
+
+  rc = mdb_dbi_open(txn, NULL, 0, &dbi);
+  if (MDB_NOTFOUND == rc) {
+
+    /* brand new database, skip the get */
+
+  } else if (rc) {
+
+    /* error happened */
+
+    ERROR(PLUGIN_NAME " plugin: mdb_dbi_open returned: %s (%d)",
+          mdb_strerror(rc), rc);
+    mdb_txn_abort(txn);
+    return -1;
+
+  } else {
+
+    /* existing database, do the get */
+
+    rc = mdb_get(txn, dbi, &key, &data);
+  }
+
+  int add = 0;
+
+  if (MDB_NOTFOUND == rc) {
+
+    /* new host found, send a notification */
+
+    add = 1;
+
+  } else if (rc) {
+
+    /* error happened */
+
+    ERROR(PLUGIN_NAME " plugin: mdb_get returned: %s (%d)", mdb_strerror(rc),
+          rc);
+    mdb_dbi_close(env, dbi);
+    mdb_txn_abort(txn);
+    return -1;
+
+  } else {
+
+    /* have we seen host in the same second as we did our last put? if so
+     * it is good enough, exit cheaply
+     */
+
+    struct timeval *tv_then = data.mv_data;
+
+    if (data.mv_size == sizeof(struct timespec) && ((tv_then = data.mv_data)) &&
```

Comment:
You don’t really need that kind of accuracy.
Common solution is update at 1/2 of threshold interval.

If your check threshold is 60 seconds, it's enough to update once per 30 seconds.

So, you can reduce locking/writes 5 times for your "10 seconds" configuration (update LMDB for each host only once per 5 seconds), 30 times for "60 seconds" configuration and 150 times for configuration with threshold of 300 sec, etc.

It is unfortunate that you ignored this comment in the comments.

Reply:
The network guys who wanted this disagreed.

Your suggestion of a CheckInterval option below resolves this - it will allow people who need accuracy to get it, while allowing people who don't want it to not.

Implemented as below.